generation/remap.rsp

```diff
@@ -0,0 +1,55 @@
+--remap
+AudioServiceHandle*=IntPtr
```

Comment:
@capnkenny, I don't like that this is `AudioServiceHandle*`. It implies, to me, that this is a pointer to a handle rather than being a handle itself.

A better name might be `NrtAudioService_t*`, etc. This makes it so you have the "friendly name" as `NrtAudioService` and then the backing hidden struct `NrtAudioService_t`, so bindings (like C#) see `NrtAudioservice_t*`. This matches what things like Vulkan do themselves.

Alternatively, this could be a proper:
```cpp
typedef struct
{
    struct NrtAudioService_t* Handle;
} NrtAudioService;
```

The benefit then is that it isn't a typedef but rather a "proper" wrapper type, which makes it simpler for bindings to just use `NrtAudioService` directly and to "ignore" the underlying handle.

Reply:
we can't use the _t suffix as that's reserved. It's why we're not using it.

Reply:
We could use `_T` like Vulkan does: https://github.com/KhronosGroup/Vulkan-Headers/blob/master/include/vulkan/vulkan_core.h#L25? There is also `_type`, `HWND__` like Windows does, and several other ways it could be disambiguated.

It's also notably not reserved by C or C++, it is only reserved by POSIX (and therefore "Unix", but I also am skeptical that they would release a type called `NrtAudioService_t` 😉).

![image](https://user-images.githubusercontent.com/10487869/112508921-3bc9a100-8d4d-11eb-9047-c9d479fd401d.png)
![image](https://user-images.githubusercontent.com/10487869/112509152-6fa4c680-8d4d-11eb-97b3-59a33ab0988e.png)

Reply:
Sure, I just wonder how much it is worth it - also you might want to talk with Monica about that since she explicitly asked us not to do this (and it took us 7 months to get to this point)

Reply:
cc @FiniteReality

Reply:
Not to use `*_t` or not to use some explicit suffix (e.g. `*_T`, `*_type`, `*__`, etc) to indicate that the underlying thing is an opaque type and not a handle?

Reply:
not to use `_t` - I don't mind changing naming conventions if you want to do so, however it is low priority for me because:
- I have graphics work to be doing at the moment myself, and Kenny is working on Audio
- It took us 7 months to even get the C API stable so I'm quite burnt out after 6 rewrites

Reply:
I'm also fine with not using `_t`, provided one of the other options is picked. Was just calling out that the term `NrtAudioServiceHandle*` will be confusing to external languages as it implies its a pointer to a handle, where in practice the pointer itself is the handle and the backing type is "opaque".

Reply:
I was saying to not use `_t` prefixes, as they are reserved by POSIX. In my interop code for animation, I've done this:

```c
typedef struct KeyframeSequenceHandle* NrtKeyframeSequence;
typedef float (*NrtEasingFunc)(float time, float begin, float change, float duration, void* userdata);
```

And that lets me do this:
```cpp
struct KeyframeSequence {
    float current_time;
    float last_value;
    size_t current_index;

    std::vector<Keyframe> keyframes;
};

extern "C" {

NrtKeyframeSequence Nrt_KeyframeSequence_create()
{
    return reinterpret_cast<NrtKeyframeSequence>(new KeyframeSequence());
}

// etc.
}
```

The name of the struct in the typedef isn't super important; since it's opaque and we're using it for its unique identity in the pointer, to avoid things like passing an NrtAnimationStateMachine to a function expecting an NrtKeyframeSequence

Reply:
> The name of the struct in the typedef isn't super important; since it's opaque and we're using it for its unique identity in the pointer, to avoid things like passing an NrtAnimationStateMachine to a function expecting an NrtKeyframeSequence

I agree with the latter half but not the first. The name is important because its what other systems like C# will end up seeing and exposing by default since they don't have typedefs and since struct wrappers aren't equivalent.

Languages like C# have basically three options:
* use `void*`
* use `IntPtr`
* use `NrtAudioServiceHandle*`

The name of the last implies its a pointer to a handle, when in fact the names are effectively reversed. That is in `typedef struct KeyframeSequenceHandle* NrtKeyframeSequence;`, the RHS (`NrtKeyframeSequence`) is the handle, while the LHS (`KeyframeSequenceHandle*`) is the underlying struct pointer and so seeing `typedef struct NrtKeyframeSequence* NrtKeyframeSequenceHandle;` makes "more" sense to me. Most other libraries do something like `typedef struct NrtKeyframeSequence_T* NrtKeyframeSequence;` however, often via some `DECLARE_HANDLE(NrtKeyFrameSequence)` helper.

Reply:
In that case, we should probably just do `typedef struct KeyframeSequence* NrtKeyframeSequence`, since as long as there's no `struct KeyframeSequence { ... };` in the header files, C code will see an opaque type

Reply:
I think we should continue prefixing all the types with `Nrt` to avoid ambiguity. The `_T` isn't "new" and is used by other libraries. Since it is uppercase, it is also POSIX compatible.

Reply:
I'm fine with `typedef struct NrtKeyframeSequence* NrtKeyframeSequence;` - that makes both of these declarations valid:

```c
void doThings()
{
    struct NrtKeyframeSequence* sequence; // referring to the tag of the struct
    NrtKeyframeSequence sequence2; // referring to the type name
}
```

Reply:
Would this change need to be done prior to this being merged? I'm not sure if we'd want to handle this in the same or a separate PR (my gut tells me it'd be a separate one that would block this from being merged)

Reply:
No, it would be a separate PR and the bindings would simply be regenerated after that goes in.